src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBase.DragDrop.cs

```diff
@@ -0,0 +1,324 @@
+
+					// For safety only, avoids double subscription
+					that.DragEnter -= OnReorderUpdated;
+					that.DragOver -= OnReorderUpdated;
+					that.DragLeave -= OnReorderCompleted;
+					that.Drop -= OnReorderCompleted;
+
+					that.DragEnter += OnReorderUpdated;
+					that.DragOver += OnReorderUpdated;
+					that.DragLeave += OnReorderCompleted;
+					that.Drop += OnReorderCompleted;
+				}
+			}
+		}
+
+		private static void OnItemContainerDragCompleted(UIElement sender, DropCompletedEventArgs innerArgs)
+		{
+			// Note: It's not the responsibility of the ListView to remove item from the source, not matter the AcceptedOperation.
+
+			if (ItemsControlFromItemContainer(sender) is ListViewBase that)
+			{
+				that.DragEnter -= OnReorderUpdated;
+				that.DragOver -= OnReorderUpdated;
+				that.DragLeave -= OnReorderCompleted;
+				that.Drop -= OnReorderCompleted;
+
+				if (that.CanDragItems)
+				{
+					var items = innerArgs.Info.Data.FindRawData(DragItemsFormatId) as IReadOnlyList<object> ?? new List<object>(0);
+					var args = new DragItemsCompletedEventArgs(innerArgs, items);
+
+					that.DragItemsCompleted?.Invoke(that, args);
+				}
+			}
+		}
+
+		private static void OnReorderUpdated(object sender, _DragEventArgs dragEventArgs)
+		{
+			var that = sender as ListView;
+			var src = dragEventArgs.DataView.FindRawData(ReorderOwnerFormatId) as ListView;
+			var item = dragEventArgs.DataView.FindRawData(ReorderItemFormatId);
+			var container = dragEventArgs.DataView.FindRawData(ReorderContainerFormatId) as FrameworkElement; // TODO: This might have changed/been recycled if scrolled 
+			if (that is null || src is null || item is null || container is null || src != that)
+			{
+				dragEventArgs.Log().Warn("Invalid reorder event.");
+
+				return;
+			}
+
+			that.UpdateReordering(dragEventArgs.GetPosition(that), container, item);
+		}
+
+		private static void OnReorderCompleted(object sender, _DragEventArgs dragEventArgs)
+		{
+			var that = sender as ListView;
+			var src = dragEventArgs.DataView.FindRawData(ReorderOwnerFormatId) as ListView;
+			var item = dragEventArgs.DataView.FindRawData(ReorderItemFormatId);
+			var container = dragEventArgs.DataView.FindRawData(ReorderContainerFormatId) as FrameworkElement; // TODO: This might have changed/been recycled if scrolled 
+			if (that is null || src is null || item is null || container is null || src != that)
+			{
+				dragEventArgs.Log().Warn("Invalid reorder event.");
+
+				return;
+			}
+
+			var updatedIndex = default(Uno.UI.IndexPath?);
+			that.CompleteReordering(container, item, ref updatedIndex);
+
+			if (that.IsGrouping
+				|| !updatedIndex.HasValue
+				|| !(dragEventArgs.DataView.FindRawData(DragItemsFormatId) is List<object> movedItems))
+			{
+				return;
+			}
+
+			var unwrappedSource = that.UnwrapItemsSource();
+			switch (unwrappedSource)
+			{
+				case null: // The ListView was created with items defined in XAML
```

Comment:
There's also the `GetItems()` method, which already returns `Items` if the unwrapped `ItemsSource` is null

Reply:
Yes but `ItemCollection` do not share a common interface (`IList` or untyped `IObservableVector`) so it's just easier to just explicitly handle that case. (It's pretty easy to make the `ItemCollection` implement the untyped `IObservableVector`, but I prefer not doing that at that point as it could cause some weird down casting checks ... I've done it in local and will validate a bit more before merging it).